File: iot/java/iotcs/csl/javase/src/main/java/com/oracle/iot/client/DeviceModelAction.java
/*
 * Copyright (c) 2015, 2016, Oracle and/or its affiliates.  All rights reserved.
 *
 * This software is dual-licensed to you under the MIT License (MIT) and 
 * the Universal Permissive License (UPL).  See the LICENSE file in the root
 * directory for license terms.  You may choose either license, or both.
 */

package com.oracle.iot.client;

import oracle.iot.client.DeviceModel;

/**
 * DeviceModelAction is the model of an action in a {@link DeviceModel}.
 */
public abstract class DeviceModelAction {

    /**
     * Get the action name.
     * @return the action name from the device model
     */
    public abstract String getName();

    /**
     * A human friendly description of the action. If the model does not
     * contain a description, this method will return an empty String.
     * @return the action description, or an empty string
     */
    public abstract String getDescription();

    /**
     * The data type of the argument to the action. If the action does not
     * take an argument, then this method will return null.
     * @return the action argument's data type, or null
     */
    public abstract DeviceModelAttribute.Type getArgType();


    /**
     * For {@link DeviceModelAttribute.Type#NUMBER} and
     * {@link DeviceModelAttribute.Type#INTEGER} only, give the lower bound of the
     * acceptable range of values for the action's argument.
     * Null is always returned for actions other than {@code NUMBER}
     * or {@code INTEGER} type.
     *
     * @return a Number, or null if no lower bound has been set
     */
    public abstract Number getLowerBound();

    /**
     * For {@link DeviceModelAttribute.Type#NUMBER} and
     * {@link DeviceModelAttribute.Type#INTEGER} only, give the upper bound of the
     * acceptable range of values for the action's argument.
     * Null is always returned for actions other than {@code NUMBER}
     * or {@code INTEGER} type.
     *
     * @return a Number, or null if no upper bound has been set
     */
    public abstract Number getUpperBound();

    /**
     * @return an alternative name for the attribute
     * @deprecated Use {@link #getName()}
     */
    @Deprecated
    public abstract String getAlias();

    protected DeviceModelAction() {}

}

File: iot/iot/csl/js/src/device/DeviceModelAction.js
/*
 * Copyright (c) 2018, Oracle and/or its affiliates.  All rights reserved.
 *
 * This software is dual-licensed to you under the MIT License (MIT) and
 * the Universal Permissive License (UPL).  See the LICENSE file in the root
 * directory for license terms.  You may choose either license, or both.
 */

/**
 * DeviceModelAction
 */
class DeviceModelAction {
    /**
     *
     * @param {string} name
     * @param {string} description
     * @param {DeviceModelAttribute.Type} type
     * @param {number} lowerBound
     * @param {number} upperBound
     * @param {string} alias
     */
    constructor(name, description, type, lowerBound, upperBound, alias) {
        // @type {string}
        this.alias = alias;
        // @type {DeviceModelAttributeType}
        this.argType = type;
        // @type {string}
        this.description = description;
        // @type {string}
        this.name = name;

        // @type {number}
        this.lowerBound = null;
        // @type {number}
        this.upperBound = null;

        if (this.argType === DeviceModelAttribute.Type.INTEGER ||
            this.argType === DeviceModelAttribute.Type.NUMBER)
        {
            this.lowerBound = lowerBound;
            this.upperBound = upperBound;
        }

        this.alias = alias;
    }


    /**
     *
     * @return {string}
     */
    getAlias() {
        return this.alias;
    }

    /**
     * @return {string} ({DeviceModelAttribute.Type})
     */
    getArgType() {
        return this.argType;
    }

    /**
     * @return {string}
     */
    getDescription() {
        return this.description;
    }

    /**
     * @return {number}
     */
    getLowerBound() {
        return this.lowerBound;
    }

    /**
     * @return {string} name
     */
    getName() {
        return this.name;
    }

    /**
     * @return {number}
     */
    getUpperBound() {
        return this.upperBound;
    }

    /**
     *
     * @return {string}
     */
    toString() {
        return 'name = ' + this.name +
            ', description = ' + this.description +
            ', type = ' + this.argType +
            ', lowerBound = ' + this.lowerBound +
            ', upperBound = ' + this.upperBound +
            ', alias = ' + this.alias;
    }
}
